### PR TITLE
JAMES-3416 ElasticSearch indexing should match full email address

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java [hunk 1]
```diff
@@ -282,7 +281,8 @@ private QueryBuilder manageAddressFields(String headerName, String value) {
             boolQuery()
                 .should(matchQuery(getFieldNameFromHeaderName(headerName) + "." + JsonMessageConstants.EMailer.NAME, value))
                 .should(matchQuery(getFieldNameFromHeaderName(headerName) + "." + JsonMessageConstants.EMailer.ADDRESS, value))
-                .should(matchQuery(getFieldNameFromHeaderName(headerName) + "." + JsonMessageConstants.EMailer.ADDRESS + "." + RAW, value)),
+            //    .should(matchQuery(getFieldNameFromHeaderName(headerName) + "." + JsonMessageConstants.EMailer.ADDRESS + "." + RAW, value))
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java:284]: Remove it?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java [hunk 2]
```diff
@@ -282,7 +281,8 @@ private QueryBuilder manageAddressFields(String headerName, String value) {
...
+            ,
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java:285]: Unneeded line break?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 3]
```diff
@@ -852,18 +852,6 @@ void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithOnlyUserPa
             .containsOnly(m8.getUid());
     }
 
-    @Test
-    void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithDomainPartOfEmail() throws Exception {
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java:856]: Shouldn't we just `@Disabled("JAMES-3406 Partial address matches are not supported")` ?

Same thing for all tests here.

File: server/protocols/jmap-rfc-8621-integration-tests/distributed-jmap-rfc-8621-integration-tests/src/test/java/org/apache/james/jmap/rfc8621/distributed/DistributedEmailQueryMethodTest.java [hunk 4]
```diff
@@ -52,4 +55,24 @@
         .server(configuration -> CassandraRabbitMQJamesServerMain.createServer(configuration)
             .overrideWith(new TestJMAPServerModule()))
         .build();
+
+    @Test
+    @Override
+    @Disabled("JAMES-XXXX Exact match for address fields is more prioritized. Partial matching is not supported currently")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/distributed-jmap-rfc-8621-integration-tests/src/test/java/org/apache/james/jmap/rfc8621/distributed/DistributedEmailQueryMethodTest.java:61]: JAMES-XXXX => JAMES-3416

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
- Can we have a changelog entry for this fix ?

 - I understand this fix is backward compatible, but I have to upgrade the ElasticSearch mapping to benefit from it. Could you mention it, and point to how to do that in the upgrade instructions?

**Comment by chibenwa** [comment]
```
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR] Failures: 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.emailWithAttachmentsShouldConvertAttachmentsWhenIndexAttachmentsIsTrue:261 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"btellier@linagora.com"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.emailWithAttachmentsShouldNotConvertAttachmentsWhenIndexAttachmentsIsFalse:289 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"btellier@linagora.com"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.emailWithNoInternalDateShouldUseNowDate:233 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"btellier@linagora.com"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.htmlEmailShouldBeWellConvertedToJson:151 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"mister@james.org"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.pgpSignedEmailShouldBeWellConvertedToJson:171 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"debian-security-announce@lists.debian.org"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.recursiveEmailShouldBeWellConvertedToJson:212 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"btellier@linagora.com"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.spamEmailShouldBeWellConvertedToJson:109 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"root@listes.minet.net"> but was: <null>.
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] 
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] [ERROR]   MessageToElasticSearchJsonTest.spamEmailShouldBeWellConvertedToJsonWithApacheTika:361 JSON documents are different:
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found when comparing expected array element to[0] to actual element to[0].
[1a996bf7e41d925594c5c4d3ea60646af4ba2863] Different value found in node "to[0].name", expected: <"btellier@linagora.com"> but was: <null>.
```

This test class (and the json files it relies on) needs to be adapted.

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
This comment is not addressed @LanKhuat

File: upgrade-instructions.md [hunk 5]
```diff
@@ -21,6 +21,24 @@ Change list:
  - [LDAP users repository connection pool now defaults to false](#ldap-users-repository-connection-pool-now-defaults-to-false)
  - [Swift support has been dropped](#swift-support-has-been-dropped)
 
+### ElasticSearch indexing should match full email address
+
+JIRA: https://issues.apache.org/jira/browse/JAMES-3416
+
+Concerned product: Distributed James
+
+Previously, we want address fields (From, To, Cc...) to be dynamically matched via ElasticSearch indexing configuration.
```
Comment (chibenwa) [inline upgrade-instructions.md:30]: Previously, address fields (From, To, Cc...) were badly tokenized, resulting in a poor search experience. For instance `bob@domain.tld` would match with `alice@domain.tld`.

We redefined the index to have exact matching on the address field and improve the search experience.

To benefit from these changes, you need to update the ElasticSearch index mapping. Procedure is described here: https://github.com/apache/james-project/blob/master/docs/modules/servers/pages/distributed/operate/guide.adoc#on-the-fly-elasticsearch-index-setting-update .

Here are the changes that needs to be applied to the existing mapping:


- `type` of the field should be changed from `text` to `keyword`.
- remove the `keep_mail_and_url` search analyzer.
- Add a `case_insensitive` normalizer for each of the address fields mapping.  

Note that you can ignore this, and keep using the old mapping, but your address search experience will not be optimal.

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
This PR, a subpart of this work , brings in the best of both worlds! https://github.com/linagora/james-project/pull/3875

**Comment by chibenwa** [comment]
https://github.com/linagora/james-project/pull/3875 solved that, this PR is no longer needed

**Pull request closed without merging**